pkg/sql/drop_function.go

```diff
@@ -12,24 +12,212 @@ package sql
+		mut, err := p.prepareMutateFunction(ctx, fnID)
+		if err != nil {
+			return nil, err
+		}
+		dropNode.toDrop = append(dropNode.toDrop, mut)
+	}
+
+	if len(dropNode.toDrop) == 0 {
+		return newZeroNode(nil), nil
+	}
+	// TODO(chengxiong): check if there is any backreference which requires
+	// CASCADE drop behavior. This is needed when we start allowing UDF
+	// references from other objects.
+	return dropNode, nil
 }
 
 func (n *dropFunctionNode) startExec(params runParams) error {
-	return unimplemented.NewWithIssue(83235, "drop function not supported")
+	for _, fnMutable := range n.toDrop {
+		if err := params.p.dropFunctionImpl(params.ctx, fnMutable); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 func (n *dropFunctionNode) Next(params runParams) (bool, error) { return false, nil }
 func (n *dropFunctionNode) Values() tree.Datums                 { return tree.Datums{} }
 func (n *dropFunctionNode) Close(ctx context.Context)           {}
+
+// matchFunction tries to resolve a user-defined function with the given
+// signature from current search path, only overload with exactly the same
+// arguments types is considered a match. If required==true, error is returned
+// if function not found. Error is also returned if a builtin function is
+// matched.
+func (p *planner) matchUDF(
+	ctx context.Context, fn *tree.FuncObj, required bool,
+) (*tree.QualifiedOverload, error) {
+	path := p.CurrentSearchPath()
+	fnDef, err := p.ResolveFunction(ctx, fn.FuncName.ToUnresolvedObjectName().ToUnresolvedName(), &path)
+	if err != nil {
+		if !required && errors.Is(err, tree.ErrFunctionUndefined) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	argTypes, err := fn.InputArgTypes(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	ol, err := fnDef.MatchOverload(argTypes, fn.FuncName.Schema(), &path)
+	if err != nil {
+		if !required && errors.Is(err, tree.ErrFunctionUndefined) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	if !ol.IsUDF {
+		return nil, errors.Errorf(
+			"cannot drop function %s because it is required by the database system",
+			ol.Signature(true /*Simplify*/),
+		)
+	}
+	return &ol, nil
+}
+
+func (p *planner) prepareMutateFunction(
+	ctx context.Context, fnID descpb.ID,
+) (*funcdesc.Mutable, error) {
+	mutable, err := p.Descriptors().GetMutableFunctionByID(ctx, p.Txn(), fnID, tree.ObjectLookupFlagsWithRequired())
+	if err != nil {
+		return nil, err
+	}
+	hasOwernship, err := p.HasOwnershipOnSchema(ctx, mutable.GetParentSchemaID(), mutable.GetParentID())
+	if err != nil {
+		return nil, err
+	}
+	hasOwernship, err = p.HasOwnership(ctx, mutable)
+	if err != nil {
+		return nil, err
+	}
+	if !hasOwernship {
+		return nil, errors.Errorf("must be owner of function %s", mutable.GetName())
+	}
+	return mutable, nil
+}
+
+func (p *planner) dropFunctionImpl(ctx context.Context, fnMutable *funcdesc.Mutable) error {
+	if fnMutable.Dropped() {
+		return errors.Errorf("function %q is already being dropped", fnMutable.Name)
+	}
```

Comment:
oh I have dedupe logic with `fnResolved` to make sure only drop a function once, but good point that there should be a test for it. I'll add one.

Reply:
actually found a bug that the the ids were not added to the deduping set :( fixed that and added tests for it.

Reply:
We should also write a test that if you try to drop something twice without IF exists, that it fails

Reply:
added, thanks for pointing that out. actually we probably won't hit into the error of this check at all since the function is already removed from the schema descriptor, so later statements dropping the same function would never find the function at all.

Reply:
also added a test for the case where there are two same udf signatures in different schemas, and two statements in transaction to drop with the same signature, all two functions should be dropped, postgres does the same.